### PR TITLE
ENH: Allow use of svd on empty arrays

### PR DESCRIPTION
**Pull request by eric-wieser**
part of #8654

Revived from over a year ago

cc @convexset

### CODE REVIEW

**Review by mhvk — APPROVED** [review]
Looks good to me

**Review by seberg — COMMENTED** [review]
The changes are probably good, I guess there are practically tests run now when deleting those lines, but... IIRC the svd should probably have a specific form at least when `full_matrices` is set to True, and I doubt that this is already being tested?

File: numpy/linalg/umath_linalg.c.src [hunk 1]
```diff
@@ -2771,6 +2770,15 @@ static NPY_INLINE void
                 if ('N' == params.JOBZ) {
                     delinearize_@REALTYPE@_matrix(args[1], params.S, &s_out);
                 } else {
+                    if ('A' == params.JOBZ && min_m_n == 0) {
+                        /* Lapack has betrayed us and left these uninitialized,
```
Comment (seberg) [inline numpy/linalg/umath_linalg.c.src]: NIT: comment should start on the next line.

**Comment by seberg** [comment]
Heh, that was quicker then possible ;), LGTM with the new test now.

**Comment by seberg** [comment]
Hmm, actually, I am not 100% sure right now if this goes into Lapack, I am a bit worried that things might not work for annoying lapack implementations. And the thing is on a quick look over I am not sure it does not sometimes go into lapack.

@eric-wieser are you pretty sure it is either not going in or lapack standard should error out cleanly?

**Comment by eric-wieser** [comment]
`full_matrices` is true by default. I've added a test for the identity matrix though, which was otherwise not tested for

**Comment by eric-wieser** [comment]
This does go into lapack, but it passes parameters that conform. Typically our mistakes in this area have been setting `LDA = M` rather than `LDA = max(M, 1)`, but `svd` already does this correctly.

I'd probably say put it in - there's no way for lapack to produce an invalid result here, since in the 0-size case all of the outputs are overwritten after the call anyway. If it returns an error code, then we can handle that if/when we get a report of it. And if it clobbers arbitrary memory... Then I think the authors of the lapack implementation need strong words.

**Comment by seberg** [comment]
OK, if you are certain enough that this is safe. I don't really know these lapack details well, vaguely remember that max(M, 1) thing. It also seems a bit like we are already doing this without complains in other functions, so seems fine.

**Comment by seberg** [comment]
OK, tests succeed, so will put this in. Thanks Eric!

**Pull request merged**